agent/agent_worker_test.go

```diff
@@ -399,3 +401,240 @@ func TestAgentWorker_DisconnectAfterJob_Start_Pause_Unpause(t *testing.T) {
 		t.Errorf("job.State = %q, want %q", got, want)
 	}
 }
+
+func TestAgentWorker_SetEndpointDuringRegistration(t *testing.T) {
+	// The registration request is made in clicommand.AgentStartCommand, and the response
+	// is passed into agent.NewAgentWorker(...), so we'll just test the response handling.
+	t.Parallel()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	t.Cleanup(cancel)
+
+	server := NewFakeAPIServer()
+	defer server.Close()
+	targetEndpoint := server.URL
+
+	const agentSessionToken = "alpacas"
+	agent := server.AddAgent(agentSessionToken)
+	agent.PingHandler = func() (api.Ping, error) {
+		switch agent.Pings {
+		case 0:
+			t.Log("server ping: disconnect")
+			return api.Ping{Action: "disconnect"}, nil
+		default:
+			return api.Ping{}, errors.New("too many pings")
+		}
+	}
+
+	// Create a listener without an HTTP server; hitting this would cause a network error/timeout.
+	// This is a bit neater than using a random host/port and hoping nothing is listening on it.
+	registrationServer, err := net.Listen("tcp", ":0")
+	if err != nil {
+		t.Fatalf("creating broken endpoint listener: %v", err)
+	}
+	defer registrationServer.Close()
+	registrationEndpoint := fmt.Sprintf("http://%s/", registrationServer.Addr().String())
```

Comment:
`fmt.Sprintf` understands the `String` method, and should call it automatically:

```suggestion
	registrationEndpoint := fmt.Sprintf("http://%s/", registrationServer.Addr())
```